File: web/src/wheelMath.ts
import type { WheelEntry } from "./types";

export const palette = [
  "D4756A",
  "E89450",
  "E8C44A",
  "5BAA7C",
  "5B8AC4",
  "8B6FBF",
  "BF6F96",
];

export type Segment = {
  start: number;
  end: number;
  mid: number;
};

export function buildSegments(entries: WheelEntry[], weighted: boolean): Segment[] {
  if (entries.length === 0) return [];

  if (!weighted) {
    const slice = 360 / entries.length;
    return entries.map((_, index) => {
      const start = -90 + index * slice;
      return { start, end: start + slice, mid: start + slice / 2 };
    });
  }

  const total = entries.reduce((sum, entry) => sum + Math.max(entry.weight, 0), 0);
  const safeTotal = total > 0 ? total : entries.length;
  let cursor = -90;

  return entries.map((entry) => {
    const weight = total > 0 ? Math.max(entry.weight, 0) : 1;
    const angle = (weight / safeTotal) * 360;
    const segment = { start: cursor, end: cursor + angle, mid: cursor + angle / 2 };
    cursor += angle;
    return segment;
  });
}

export function pickWinner(entries: WheelEntry[], weighted: boolean, excluding?: string): number {
  if (entries.length === 0) return 0;

  let pool = entries.map((_, index) => index).filter((index) => entries[index].label !== excluding);
  if (pool.length === 0) pool = entries.map((_, index) => index);

  if (!weighted) {
    return pool[Math.floor(Math.random() * pool.length)] ?? 0;
  }

  const total = pool.reduce((sum, index) => sum + Math.max(entries[index].weight, 0), 0);
  const target = Math.random() * (total > 0 ? total : pool.length);
  let cursor = 0;

  for (const index of pool) {
    cursor += total > 0 ? Math.max(entries[index].weight, 0) : 1;
    if (target < cursor) return index;
  }

  return pool[pool.length - 1] ?? 0;
}

export function spinDelta(segment: Segment, currentRotation: number, spins = 7): number {
  const margin = (segment.end - segment.start) * 0.15;
  const target = randomBetween(segment.start + margin, segment.end - margin);
  const mod = currentRotation % 360;
  const normalized = mod < 0 ? mod + 360 : mod;
  return spins * 360 + (-90 - target - normalized);
}

export function nextColorHex(entries: WheelEntry[]): string {
  const counts = new Map<string, number>();
  entries.forEach((entry) => counts.set(entry.colorHex, (counts.get(entry.colorHex) ?? 0) + 1));
  const previous = entries.at(-1)?.colorHex;
  return (
    palette
      .filter((color) => color !== previous)
      .sort((a, b) => (counts.get(a) ?? 0) - (counts.get(b) ?? 0))[0] ??
    palette[entries.length % palette.length]
  );
}

function randomBetween(min: number, max: number): number {
  return min + Math.random() * (max - min);
}

File: SpinFlipRoll/SpinFlipRoll 2/WheelMath.swift
// WheelMath.swift
// Add to BOTH the SpinFlipRoll (iOS) and SpinFlipRollWatch (watchOS) targets.
import Foundation

enum WheelMath {
    static let palette = [
        "E05555", // red
        "F08020", // orange
        "F0B520", // yellow
        "2FA060", // green
        "3080D0", // blue
        "8855CC", // purple
        "CC4488", // pink
    ]

    struct Segment {
        let start: Double
        let end: Double
        let mid: Double
    }

    static func buildSegments(_ entries: [SharedEntry], weighted: Bool) -> [Segment] {
        guard !entries.isEmpty else { return [] }

        if weighted {
            let total = entries.reduce(0.0) { $0 + max($1.weight, 0) }
            let safe = total > 0 ? total : Double(entries.count)
            var cursor = -90.0
            return entries.map { entry in
                let w = total > 0 ? max(entry.weight, 0) : 1.0
                let angle = (w / safe) * 360.0
                let seg = Segment(start: cursor, end: cursor + angle, mid: cursor + angle / 2)
                cursor += angle
                return seg
            }
        } else {
            let slice = 360.0 / Double(entries.count)
            return entries.indices.map { i in
                let start = -90.0 + Double(i) * slice
                return Segment(start: start, end: start + slice, mid: start + slice / 2)
            }
        }
    }

    /// Returns the index of the winner, respecting optional weighted and no-repeat logic.
    static func pickWinner(
        _ entries: [SharedEntry],
        weighted: Bool,
        excluding: String? = nil
    ) -> Int {
        guard !entries.isEmpty else { return 0 }

        var pool = entries.indices.filter { excluding == nil || entries[$0].label != excluding }
        if pool.isEmpty { pool = Array(entries.indices) }

        guard weighted else { return pool.randomElement() ?? 0 }

        let total = pool.map { max(entries[$0].weight, 0) }.reduce(0, +)
        let rand = Double.random(in: 0..<(total > 0 ? total : Double(pool.count)))
        var cum = 0.0
        for idx in pool {
            cum += total > 0 ? max(entries[idx].weight, 0) : 1.0
            if rand < cum { return idx }
        }
        return pool.last ?? 0
    }

    /// Returns how many extra degrees to rotate so the chosen segment lands under the top pointer.
    static func spinDelta(winnerMid: Double, currentRotation: Double, spins: Int = 7) -> Double {
        let mod = currentRotation.truncatingRemainder(dividingBy: 360)
        let norm = mod < 0 ? mod + 360 : mod
        return Double(spins) * 360.0 + (-90.0 - winnerMid - norm)
    }
}
